ext_proc: Implement response path for headers only

gbrail opened the pull request:
Implement header processing on the response path by sending the
response_headers message to the processor and handling the result.

Also update the docs in the .proto file.

Signed-off-by: Gregory Brail <gregbrail@google.com>

htuch commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -48,66 +50,121 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
+    encoder_callbacks_->continueEncoding();
+    return true;
+  }
+  return false;
```
Should the request/response path be factored to a common method? They look the same pretty much.

gbrail replied:
I refactored a bit in the most recent commit.

htuch commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -48,66 +50,121 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
+bool Filter::handleImmediateResponse(const ImmediateResponse& response) {
+  if (response_state_ == FilterState::HEADERS) {
+    // Waiting for a response headers response, so return immediately now.
+    // Do this first in case both are in progress.
```
As an observation, the decoupling of request/response state from server reply message type makes it a bit confusing to reason about the state machine here. If we had distinct message for "send immediate response" for request/reply, it would be a bit more intuitive in the code at least; OTOH, I think the behaviors are identical so it's technically not needed.

gbrail replied:
I'm not entirely sure how to improve this area. For this PR I switched things up to have one function that handles the incoming message and switches on the message type, and others that decide whether it's an appropriate time to handle each one. Or are you suggesting changing the proto so that "immediate_response" is part of a larger message type rather than a separate message?

htuch replied:
I guess tracking the containing context (request/response) for the immediate response would be clearer than relying on request/response state variables to me, but I can see either working.

htuch commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -48,66 +50,121 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
+    ImmediateResponse error_response;
+    error_response.mutable_status()->set_code(envoy::type::v3::StatusCode::InternalServerError);
+    error_response.set_details(absl::StrFormat("%s: gRPC error %i", kErrorPrefix, status));
+    handleImmediateResponse(error_response);
```
Is the only situation in which you can send this synthetic error response when we're in the HEADERS request/response state? I'm thinking there might be a broader set of states...

gbrail replied:
In the current codebase, "handleImmediateResponse" only needs to work in the "HEADERS" state. When we implement body processing two PRs from now, that will expand. However, there's a lot I need to experiment with regarding how Envoy handles "sendLocalReply" on both code paths before getting it right.

htuch replied:
Why do we need to wait until we are in response HEADERS to send an error message? I think as long as we haven't sent any response headers, we can do a `sendLocalReply`. We don't need to wait for `encodeHeaders` to do this. My original comment is that we should be able to do a `sendLocalReply` in any situation before we have sent the response headers back.

If you look at the behavior of something like `ConnectionManagerImpl::ActiveStream::onIdleTimeout()`, this has conditional behavior on whether we've processed response headers or not.

htuch reviewed (changes requested):
Looks good, a handful of comments..
/wait

htuch commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -48,66 +50,121 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
+    decoder_callbacks_->continueDecoding();
+    return true;
+  }
+  return false;
```
FYI there's missing coverage on the `false` path, see https://storage.googleapis.com/envoy-pr/de1bd48/coverage/source/extensions/filters/http/ext_proc/ext_proc.cc.gcov.html (otherwise the coverage is excellent!).

gbrail replied:
I added a set of new tests that just focus on out of order messages in a way that's slightly easier to expand. Let's see if that does the trick. (For some reason, I can do a coverage test run locally but the coverage HTML can't be generated locally on my machine.)

gbrail commented:
GitHub resolved it, but I don't know why I thought that sendLocalResponse couldn't be called in more places. I cleaned up the code and added an integration test.

htuch commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -48,66 +50,104 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
 
+FilterHeadersStatus Filter::encodeHeaders(ResponseHeaderMap& headers, bool end_of_stream) {
+  if (stream_closed_) {
+    return FilterHeadersStatus::Continue;
```
Should this keep forwarding on a closed management stream or halt? How come we don't reset the stream in `closeStream()`?

gbrail replied:
"closeStream" is supposed to close the stream to the external processor. We just do this by calling "close" on the gRPC stream. Is there something else that we're supposed to to?

As for forwarding on a closed management stream -- yes. The protocol allows the external processor to close the stream at any time, at which point the filter just skips it. The configuration also allows us to set "ignore errors," which means that we treat a failed stream the same as a closed stream. Finally, if there is an error on the stream, then we "sendLocalReply," but the filter still gets invoked after that in some cases so we still want to continue AFAIK.

htuch replied:
Right, my point was that if you have closed the stream to the management server (e.g. on error) but aren't doing "ignore errors", do you still want to potentially forward traffic?

gbrail replied:
Right now, this is what happens:

decodeHeaders()
error on gRPC stream
filter sets "stream_closed_" and calls sendLocalReply()

at this point, it's been my experience that sendLocalReply will call encodeHeaders, presumably to give us a chance to do more processing. We don't want to do anything else at this point since the processor stream failed so we have no choice IMO but to continue. Is there another option there?

gbrail replied:
FWIW I changed the filter to always call sendLocalReply on encoder_callbacks_, and now the extra call to encodeHeaders doesn't show up. So the only way we should get to encodeHeaders with stream_closed_ set to true is if either the remote processor closed the stream or if we closed it while ignoring an error.

htuch commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -117,34 +157,39 @@ void Filter::onGrpcClose() {
-      },
-      grpc_status, response.details());
+  if (on_decoding) {
+    decoder_callbacks_->sendLocalReply(static_cast<Http::Code>(status_code), response.body(),
```
Does it matter if you use decoder/encode callbacks here? I don't recall TBH, but this seems strange that it would make a difference.

gbrail replied:
It does seem strange but the API docs don't really say so I was trying to play it safe. Can we be sure that there will not be a difference in some future version of Envoy?

htuch replied:
These are the same inside of Envoy, there is no functional difference. The reason you have both is you have an encoder-decoder filter, whereas if you just had decoder (for example), you would have one. I think you can safely just use the decoder callbacks.

gbrail replied:
OK -- I'll change that and resolve.

htuch commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -48,66 +50,104 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of
+  closeStream();
+
+  if (encoding_started_) {
+    ENVOY_LOG(debug, "Returning immediate response from processor on encoding path");
```
Once encoding has actually started, I don't think we can do any local reply any more.

gbrail replied:
The comments in filter.h say this:

"If a response has already started (e.g. if the router calls sendSendLocalReply after encoding headers) this will either ship the reply directly to the downstream codec, or reset the stream."

This definitely works after encodeHeaders has been called and before we call continueEncoding. After that we don't have many options, so we either call this or we somehow replace the response body ourselves. When we get to implementing body handling I planned to write integration tests to make sure we understand what Envoy will actually do.

gbrail commented:
Thanks -- I answered your comments but I don't know of an actionable thing to change right now.

htuch reviewed (approved):
LGTM, thanks!

htuch commented:
@gbrail in future PRs, could you add the "Risk level", "Testing", etc. template fields? These are helpful to folks going through history trying to search for risky changes etc.
